Document how to build with Ninja.

david-german-tri opened the pull request:
Fixes #2283 

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2555)
<!-- Reviewable:end -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (platform) Please confirm that you generated the Sphinx doc and were satisfied with it. Is there a Ninja user who can feature review the instructions?

<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> #### Document how to build with Ninja\.
</details>


---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKAcKjZsYxQmGxpNOXy:-829245671)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I'm following the instructions now.

<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> :lgtm: (platform) Please confirm that you generated the Sphinx doc and were satisfied with it. Is there a Ninja user who can feature review the instructions?
</details>


---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKAd4sDwc_3mrYL8NGw:-734203041)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[drake/doc/from_source.rst, line 97 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKAdNH4cyMbf1H1m1hm:-KKAdNH51EbRWTTAYoq3:1977481322) ([raw file](https://github.com/robotlocomotion/drake/blob/42aafd7ad19ac2eb5c82389e49bc84645bbc9566/drake/doc/from_source.rst#L97)):*
> ```reStructuredText
> ::
> 
>     mkdir drake-build
> ```

I think there may be a missing `cd drake-build` after the above line.

---

*[drake/doc/from_source.rst, line 98 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKAeF88nArDNXPFaWS_:-KKAeF88nArDNXPFaWSa:1297458169) ([raw file](https://github.com/robotlocomotion/drake/blob/42aafd7ad19ac2eb5c82389e49bc84645bbc9566/drake/doc/from_source.rst#L98)):*
> ```reStructuredText
> 
>     mkdir drake-build
>     cmake path/to/drake-distro -G Ninja
> ```

I'm encountering the following error after executing the above comand:

```
$ cmake ../drake -G Ninja
-- The C compiler identification is GNU 4.9.3
-- The CXX compiler identification is GNU 4.9.3
-- Check for working C compiler using: Ninja
CMake Error:
  The detected version of Ninja (log: ninja version 0.1.3 initializing) is
  less than the version of Ninja required by CMake (1.3).


CMake Error: Internal CMake error, TryCompile generation of cmake failed
-- Check for working C compiler using: Ninja -- broken
CMake Error at /home/liang/tools/cmake-3.5.2-Linux-x86_64/share/cmake-3.5/Modules/CMakeTestCCompiler.cmake:61 (message):
  The C compiler "/usr/lib/ccache/gcc-4.9" is not able to compile a simple
  test program.

  It fails with the following output:

   

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:2 (project)


-- Configuring incomplete, errors occurred!
See also "/home/liang/dev/drake_build/CMakeFiles/CMakeOutput.log".
See also "/home/liang/dev/drake_build/CMakeFiles/CMakeError.log".
```
Help!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/doc/from_source.rst, line 97 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKAdNH4cyMbf1H1m1hm:-KKAgEfzRO06z8i1-Am8:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/42aafd7ad19ac2eb5c82389e49bc84645bbc9566/drake/doc/from_source.rst#L97)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I think there may be a missing `cd drake-build` after the above line.
</details>
Done.

---

*[drake/doc/from_source.rst, line 98 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKAeF88nArDNXPFaWS_:-KKAjt5c3zboAXKNn6jy:-933624281) ([raw file](https://github.com/robotlocomotion/drake/blob/42aafd7ad19ac2eb5c82389e49bc84645bbc9566/drake/doc/from_source.rst#L98)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I'm encountering the following error after executing the above comand:
> 
> ```
> $ cmake ../drake -G Ninja
> -- The C compiler identification is GNU 4.9.3
> -- The CXX compiler identification is GNU 4.9.3
> -- Check for working C compiler using: Ninja
> CMake Error:
>   The detected version of Ninja (log: ninja version 0.1.3 initializing) is
>   less than the version of Ninja required by CMake (1.3).
> 
> 
> CMake Error: Internal CMake error, TryCompile generation of cmake failed
> -- Check for working C compiler using: Ninja -- broken
> CMake Error at /home/liang/tools/cmake-3.5.2-Linux-x86_64/share/cmake-3.5/Modules/CMakeTestCCompiler.cmake:61 (message):
>   The C compiler "/usr/lib/ccache/gcc-4.9" is not able to compile a simple
>   test program.
> 
>   It fails with the following output:
> 
>    
> 
>   
> 
>   CMake will not be able to correctly generate this project.
> Call Stack (most recent call first):
>   CMakeLists.txt:2 (project)
> 
> 
> -- Configuring incomplete, errors occurred!
> See also "/home/liang/dev/drake_build/CMakeFiles/CMakeOutput.log".
> See also "/home/liang/dev/drake_build/CMakeFiles/CMakeError.log".
> ```
> Help!
</details>
From offline debugging: @liangfok installed the dpkg named `ninja`, which is an unrelated piece of software.  We want `ninja-build`.  Updated Ubuntu install instructions accordingly.  Leaving this thread open until I've determined the appropriate `brew` command and updated the Mac instructions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/doc/from_source.rst, line 99 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKAnqEdaUUzNLY-W1BI:-KKAnqEdaUUzNLY-W1BJ:-521399128) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L99)):*
> ```reStructuredText
>     mkdir drake-build
>     cd drake-build
>     cmake path/to/drake-distro -G Ninja
> ```

at r2, @liangfok has reported another error:

```
$ cmake ../drake -G Ninja
-- The C compiler identification is GNU 4.9.3
-- The CXX compiler identification is GNU 4.9.3
-- Check for working C compiler using: Ninja
CMake Error:
  The detected version of Ninja () is less than the version of Ninja required
  by CMake (1.3).
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
@sherm1 Yes, I've confirmed the Sphinx output looks good.  I'm still trying to figure out why these instructions aren't working for @liangfok.  If anyone else on Ubuntu wants to give it a try, I'd be grateful.

<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I'm following the instructions now.
</details>


---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKAv0Y1MhR-A5VPJ2Gn:1078602575)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/doc/from_source.rst, line 99 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKAnqEdaUUzNLY-W1BI:-KKBoNkmkW6CLQKe80tw:1463840112) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L99)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> at r2, @liangfok has reported another error:
> 
> ```
> $ cmake ../drake -G Ninja
> -- The C compiler identification is GNU 4.9.3
> -- The CXX compiler identification is GNU 4.9.3
> -- Check for working C compiler using: Ninja
> CMake Error:
>   The detected version of Ninja () is less than the version of Ninja required
>   by CMake (1.3).
> ```
</details>
The above error occurred on my workstation, which was running Ubuntu 14.04.4. I tried to replicate the problem on my laptop using the same OS but couldn't. I was able to get past the `cmake path/to/drake-distro -G Ninja` command, but the subsequent `ninja` command resulted in the following error:

```
$ ninja
...
Submodule 'lcm/lcm-src/cmake' (https://github.com/RobotLocomotion/cmake.git) registered for path 'lcm/lcm-src/cmake'
Cloning into 'lcm/lcm-src/cmake'...
Submodule path 'lcm/lcm-src/cmake': checked out 'abe1a22d0bb767423de4e3bd1123cc2deae1c82a'
ninja: build stopped: subcommand failed.
```
I got around the above error by executing `ninja` again.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I tried it in Ubuntu though from home with no Matlab and it seems to work. I'll try tomorrow a full compilation with Matlab and I'll let you know how it went.

<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> @sherm1 Yes, I've confirmed the Sphinx output looks good.  I'm still trying to figure out why these instructions aren't working for @liangfok.  If anyone else on Ubuntu wants to give it a try, I'd be grateful.
</details>


---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/doc/from_source.rst, line 97 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBf_fjN11dawH2i6K_:-KKBf_fjN11dawH2i6Ka:-375504289) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L97)):*
> ```reStructuredText
> ::
> 
>     mkdir drake-build
> ```

the `drake-build` directory is not ignored by git. Should it be added to the `.gitignore`?

---

*[drake/doc/from_source.rst, line 99 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBcvRhCQzh0QPAK1_j:-KKBcvRhCQzh0QPAK1_k:-803028168) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L99)):*
> ```reStructuredText
>     mkdir drake-build
>     cd drake-build
>     cmake path/to/drake-distro -G Ninja
> ```

I had to do a `env CXX=g++-4.9 CC=gcc-4.9 cmake  path/to/drake-distro -G Ninja`. Maybe worth mentioning for the ones like me that need a detailed step-by-step recipe? 

---

*[drake/doc/from_source.rst, line 107 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBmaXSOd9FrEL-Fnxw:-KKBmaXSOd9FrEL-Fnxx:-741340442) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L107)):*
I am compiling from home, no Matlab. So I tried:

> `env CXX=g++-4.9 CC=gcc-4.9 cmake path/to/drake-distro -G Ninja -DDISABLE_MATLAB=ON`

That worked. Maybe worth documenting?

---

*[drake/doc/from_source.rst, line 111 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBjQ1LjEhf08zJQgvK:-KKBjQ1LjEhf08zJQgvL:-1579605925) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L111)):*
> ```reStructuredText
> 
> To determine the shell commands, compiler flags, and linker flags that CMake
> generated, consult ``build.ninja`` and ``drake/build.ninja``, or run
> ```

Not the most user friendly files I've seen. Is there an equivalent to `make options` that goes into the ccmake interface?

---

*[drake/doc/from_source.rst, line 113 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBmco8jSNbn47f1Vtk:-KKBmco8jSNbn47f1Vtl:-736275578) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L113)):*
> ```reStructuredText
> generated, consult ``build.ninja`` and ``drake/build.ninja``, or run
> ``ninja -v`` for a verbose build.
> ```

Other questions:

- How do I build from scratch at the drake level? (without building at the superbuild level).
- I also noticed that from within `drake-distro/drake-build/drake` I can build specific targets with `ninja -j <num_CPUs> target_name`. Worth documenting?
- The progress is reported on a per-target basis. I can see a list of files being compiled with `ninja -v`. However I get a super lengthy chunk of information with all the compilation flags and paths. Is there a way to ask for a reporting level somewhere in between?  
- Out-of-source questions:
    - Could I have a `drake-build/Release` and a `drake-build/Debug`? is that the recommended usage?
    - Sometimes compilation gets slow when checking out different branches. Would you recommend having a `drake-build/branch_name/[Release|Debug]`?

---

*[drake/doc/from_source.rst, line 124 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBbpk2oNzkSEuVZEn3:-KKBbpk3I6yghTg2MNxo:-647053299) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L124)):*
> ```reStructuredText
> 
>     make --version
> ```

Probably people like @liangfok and me that are switching from make to Ninja might need to install ninja. Of course we won't go to the Ubuntu install instructions to find out the appropriate ninja-build package. Would it be a good idea to reference that prerequisite here? just a link to the updated Ubuntu install instructions should do it I think.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKBp09Iq5mFnnKrujb0:-1560825941)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/doc/from_source.rst, line 97 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBf_fjN11dawH2i6K_:-KKBq7SNBYqyyMrZ8DS4:191004483) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L97)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> the `drake-build` directory is not ignored by git. Should it be added to the `.gitignore`?
</details>
I believe `drake-build` should not be located within `drake-distro`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Between, this looks great! thank you for documenting it and educating us! :-)

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I tried it in Ubuntu though from home with no Matlab and it seems to work. I'll try tomorrow a full compilation with Matlab and I'll let you know how it went.
</details>


---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKBq2n0x3aSQ1pQr1Rw:-707203207)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/doc/from_source.rst, line 99 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKAnqEdaUUzNLY-W1BI:-KKDjFo_5Bpe6Mgt0PrQ:222803957) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L99)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> The above error occurred on my workstation, which was running Ubuntu 14.04.4. I tried to replicate the problem on my laptop using the same OS but couldn't. I was able to get past the `cmake path/to/drake-distro -G Ninja` command, but the subsequent `ninja` command resulted in the following error:
> 
> ```
> $ ninja
> ...
> Submodule 'lcm/lcm-src/cmake' (https://github.com/RobotLocomotion/cmake.git) registered for path 'lcm/lcm-src/cmake'
> Cloning into 'lcm/lcm-src/cmake'...
> Submodule path 'lcm/lcm-src/cmake': checked out 'abe1a22d0bb767423de4e3bd1123cc2deae1c82a'
> ninja: build stopped: subcommand failed.
> ```
> I got around the above error by executing `ninja` again.
</details>
Just following up. On my laptop running Ubuntu 14.04.4, after running `ninja` again, I ran into the above error again, just in a different place. I then created a new build artifacts directory and tried again, this time disabling Matlab, and it worked:

```
$ mkdir build-artifacts-directory
$ cd build-artifacts-directory
$ env CXX=g++-4.9 CC=gcc-4.9 cmake ../drake-distro -G Ninja -DDISABLE_MATLAB=ON
$ ninja
```
Thus, the above error may be Matlab related. I'm not certain because the error might be intermittent and I may have just gotten lucky.

I haven't figured out why my workstation thinks the version of Ninja is wrong. I'll look more into that later this afternoon.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/doc/from_source.rst, line 76 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKDkp_V8E0PxV5prhWh:-KKDkp_V8E0PxV5prhWi:-142620728) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L76)):*
> ```reStructuredText
> There are two supported ways to build Drake:
> `Ninja <https://ninja-build.org/>`_ (recommended) and
> `Make <https://www.gnu.org/software/make/>`_ (legacy).
> ```

Can we add a brief comment or link indicating the pros / cons of `ninja` versus `make`? Is it just that `ninja` is faster but has less features relative to `make`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I tested building with Matlab in Ubuntu. It worked for me.

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKEWiKzC6op5yJsefpp:1572457218)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I'm back in CAM1 now and will look into debugging my workstation soon.

<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I tested building with Matlab in Ubuntu. It worked for me.
</details>


---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKEfuuX1-QkvckHqMBQ:578180200)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Note that on CI we have always used the latest version of Ninja (1.7.1). I would strongly recommend using that instead of whatever Ubuntu provides. Installation is just extracting an archive.

liangfok commented:
On OS X, I installed `ninja` by executing:

```
$ brew install ninja
$ ninja --version
1.6.0
```
I believe the above command needs to be included in http://drake.mit.edu/homebrew.html.

Unfortunately, the actual build using the default build flags failed with the following error:

```
$ cmake ../drake-distro -G Ninja
$ ninja
...
[153/572] Linking CXX shared module /Users/liang/dev/drake/drake/util/barycentricInterpolation.mexmaci64
FAILED: : && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -Werror=all -Wno-missing-braces -Wno-unused-private-field -DGTEST_DONT_DEFINE_FAIL=1 -DGTEST_DONT_DEFINE_SUCCEED=1 -DGTEST_DONT_DEFINE_TEST=1  -bundle -Wl,-headerpad_max_install_names  -o /Users/liang/dev/drake/drake/util/barycentricInterpolation.mexmaci64 util/CMakeFiles/barycentricInterpolation.dir/barycentricInterpolation.cpp.o -L/Users/liang/dev/drake-build/drake/lib  -L/Users/liang/dev/drake-build/install/lib  && :
Undefined symbols for architecture x86_64:
  "_mexErrMsgIdAndTxt", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mexPrintf", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxCreateDoubleMatrix_730", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxCreateNumericMatrix_730", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxGetCell_730", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxGetData", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxGetM", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxGetN", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxGetNumberOfElements", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxGetPr", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
FAILED: : && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -Werror=all -Wno-missing-braces -Wno-unused-private-field -DGTEST_DONT_DEFINE_FAIL=1 -DGTEST_DONT_DEFINE_SUCCEED=1 -DGTEST_DONT_DEFINE_TEST=1  -bundle -Wl,-headerpad_max_install_names  -o /Users/liang/dev/drake/drake/util/publishLCMLog.mexmaci64 util/CMakeFiles/publishLCMLog.dir/publishLCMLog.cpp.o -L/Users/liang/dev/drake-build/drake/lib  -L/Users/liang/dev/drake-build/install/lib -L/Users/liang/dev/drake-build/install/lib -L/usr/local/Cellar/glib/2.46.2/lib -L/usr/local/opt/gettext/lib -llcm -lgthread-2.0 -lglib-2.0 -lintl -Wl,-rpath,/Users/liang/dev/drake-build/drake/lib -Wl,-rpath,/Users/liang/dev/drake-build/install/lib && :
Undefined symbols for architecture x86_64:
  "_mexErrMsgTxt", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mexPrintf", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mxArrayToString", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mxFree", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mxGetData", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mxGetFieldByNumber_730", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mxGetFieldNumber", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mxGetNumberOfElements", referenced from:
      _mexFunction in publishLCMLog.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
warning: [options] bootstrap class path not set in conjunction with -source 1.7
1 warning
warning: [options] bootstrap class path not set in conjunction with -source 1.7
1 warning
ninja: build stopped: subcommand failed.
```

Unfortunately, I ran into the same problem after disabling Matlab:

```
$ cmake ../drake-distro -G Ninja -DDISABLE_MATLAB=ON
$ ninja
...
[14/392] Linking CXX shared module /Users/liang/dev/drake/drake/util/barycentricInterpolation.mexmaci64
FAILED: : && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -Werror=all -Wno-missing-braces -Wno-unused-private-field -DGTEST_DONT_DEFINE_FAIL=1 -DGTEST_DONT_DEFINE_SUCCEED=1 -DGTEST_DONT_DEFINE_TEST=1  -bundle -Wl,-headerpad_max_install_names  -o /Users/liang/dev/drake/drake/util/publishLCMLog.mexmaci64 util/CMakeFiles/publishLCMLog.dir/publishLCMLog.cpp.o -L/Users/liang/dev/drake-build/drake/lib  -L/Users/liang/dev/drake-build/install/lib -L/Users/liang/dev/drake-build/install/lib -L/usr/local/Cellar/glib/2.46.2/lib -L/usr/local/opt/gettext/lib -llcm -lgthread-2.0 -lglib-2.0 -lintl -Wl,-rpath,/Users/liang/dev/drake-build/drake/lib -Wl,-rpath,/Users/liang/dev/drake-build/install/lib && :
Undefined symbols for architecture x86_64:
  "_mexErrMsgTxt", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mexPrintf", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mxArrayToString", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mxFree", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mxGetData", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mxGetFieldByNumber_730", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mxGetFieldNumber", referenced from:
      _mexFunction in publishLCMLog.cpp.o
  "_mxGetNumberOfElements", referenced from:
      _mexFunction in publishLCMLog.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
FAILED: : && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -Werror=all -Wno-missing-braces -Wno-unused-private-field -DGTEST_DONT_DEFINE_FAIL=1 -DGTEST_DONT_DEFINE_SUCCEED=1 -DGTEST_DONT_DEFINE_TEST=1  -bundle -Wl,-headerpad_max_install_names  -o /Users/liang/dev/drake/drake/util/barycentricInterpolation.mexmaci64 util/CMakeFiles/barycentricInterpolation.dir/barycentricInterpolation.cpp.o -L/Users/liang/dev/drake-build/drake/lib  -L/Users/liang/dev/drake-build/install/lib  && :
Undefined symbols for architecture x86_64:
  "_mexErrMsgIdAndTxt", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mexPrintf", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxCreateDoubleMatrix_730", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxCreateNumericMatrix_730", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxGetCell_730", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxGetData", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxGetM", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxGetN", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxGetNumberOfElements", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
  "_mxGetPr", referenced from:
      _mexFunction in barycentricInterpolation.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
warning: [options] bootstrap class path not set in conjunction with -source 1.7
1 warning
warning: [options] bootstrap class path not set in conjunction with -source 1.7
1 warning
ninja: build stopped: subcommand failed.
ninja: build stopped: subcommand failed.
```
I just saw @jamiesnape's comment about using the latest version of ninja and will try upgrading now.

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/doc/from_source.rst, line 99 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKAnqEdaUUzNLY-W1BI:-KKEjfZdi8RxUpSMyXlg:260834676) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L99)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Just following up. On my laptop running Ubuntu 14.04.4, after running `ninja` again, I ran into the above error again, just in a different place. I then created a new build artifacts directory and tried again, this time disabling Matlab, and it worked:
> 
> ```
> $ mkdir build-artifacts-directory
> $ cd build-artifacts-directory
> $ env CXX=g++-4.9 CC=gcc-4.9 cmake ../drake-distro -G Ninja -DDISABLE_MATLAB=ON
> $ ninja
> ```
> Thus, the above error may be Matlab related. I'm not certain because the error might be intermittent and I may have just gotten lucky.
> 
> I haven't figured out why my workstation thinks the version of Ninja is wrong. I'll look more into that later this afternoon.
> 
</details>
After rebooting my workstation, I was unable to replicate the problem where the detected version of Ninja was less than the required version. The build worked using both`-DDISABLE_MATLAB=ON` and `-DDISABLE_MATLAB=OFF`. Because it worked, I am marking this thread as "satisfied".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKEkhjonb1ql_bBT0Ez:351922820)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I upgraded to Ninja 1.7 on my OS X machine and ran into the following error when building Drake:

```
$ ninja --version
1.7.1
$ cmake ../drake -G Ninja
$ ninja
...
[100%] Completed 'director'
[100%] Built target director
ninja: build stopped: subcommand failed.
```
Unfortunately, there was so much text printed to the terminal I couldn't find the actual error. It's possible the actual error was pushed out of the terminal's history buffer.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKF0lN_ILIOHiXnS2ie:642576888)*
<!-- Sent from Reviewable.io -->

jamiesnape commented on drake/doc/from_source.rst:
```diff
@@ -71,16 +71,67 @@ Before running build, you will need to follow the instructions for your host sys
+::
+
+    cmake path/to/drake-distro -G Ninja -DCMAKE_BUILD_TYPE:STRING=[Debug|Release|RelWithDebInfo]
+    ninja -j <num_CPUs>
```
Ninja defaults to parallel builds, so unless you wish to limit the number of cores, you can safely omit `-j`.

amcastro-tri commented:
@liangfok: I am using Ninja 1.3.4 and it seems to work. Though I had to do `env CXX=g++-4.9 CC=gcc-4.9 cmake path/to/drake-distro -G Ninja`. When I had the "no Matlab problem" at home, I redirected the output to a file like so `ninja -j 1 1>log 2>err` (I used -j 1 though I think that's the default?). Then I searched through the log file until I found the license problem. Your problem above seems to say some mex functions (Matlab!) are not defined. Maybe the compilation crashed at some point before that report and you don't get to see it? it might still be a license problem (search for "license" in the log file).

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKF2I9ZI9WNKH2tilJE:-1145165846)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/doc/from_source.rst, line 108 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKA_ygR4x6XdAinJhEE-r2-108:-KKF3ObFgCoqPp_65yB2:356033942) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L108)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary>
> Ninja defaults to parallel builds, so unless you wish to limit the number of cores, you can safely omit `-j`.
</details>
wow! look at this @liangfok! maybe mine worked because I explicitly used `-j 1`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Here is the error that causes the build to fail:

```
FAILED: eigen-prefix/src/eigen-stamp/eigen-update
cd /Users/liang/dev/drake/externals/eigen && /usr/local/Cellar/cmake/3.5.0/bin/cmake -E echo_append && /usr/local/Cellar/cmake/3.5.0/bin/cmake -E touch /Users/liang/dev/drake-build/eigen-prefix/src/eigen-stamp/eigen-update
/bin/sh: line 0: cd: /Users/liang/dev/drake/externals/eigen: No such file or directory
```

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/doc/from_source.rst, line 108 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKA_ygR4x6XdAinJhEE-r2-108:-KKFK_x1PO1d6qn5MO7g:-1992875272) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L108)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> wow! look at this @liangfok! maybe mine worked because I explicitly used `-j 1`?
</details>
Unfortunately, using `-j 1` did not fix the problem.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKFKqoUcdsWkMzteHND:-874650812)*
<!-- Sent from Reviewable.io -->

bradking commented:
> /bin/sh: line 0: cd: /Users/liang/dev/drake/externals/eigen: No such file or directory

This is due to a missing dependency in the Ninja build system caused by the use of two separate `ExternalProject_Add` calls that share a source tree.  PR #2487 should resolve this because it re-combines the calls into a single `ExternalProject_Add` call per external source tree.

jamiesnape commented:
@drake-jenkins-bot retest this please

jamiesnape commented:
@drake-jenkins-bot retest this please

david-german-tri commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/doc/from_source.rst, line 76 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKDkp_V8E0PxV5prhWh:-KKGD8BwKHUB0ZXQc46P:1371943443) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L76)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Can we add a brief comment or link indicating the pros / cons of `ninja` versus `make`? Is it just that `ninja` is faster but has less features relative to `make`?
</details>
I think the linked Ninja page does a fairly good job explaining itself.  I don't have anything to add.

---

*[drake/doc/from_source.rst, line 97 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBf_fjN11dawH2i6K_:-KKGBWabsfazrWgRrUci:-1905827980) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L97)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I believe `drake-build` should not be located within `drake-distro`.
</details>
As @liangfok said and is noted in the paragraph above, the build is out-of-source.

---

*[drake/doc/from_source.rst, line 99 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBcvRhCQzh0QPAK1_j:-KKGBncBZnE-5I3j_8fK:1741247307) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L99)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I had to do a `env CXX=g++-4.9 CC=gcc-4.9 cmake  path/to/drake-distro -G Ninja`. Maybe worth mentioning for the ones like me that need a detailed step-by-step recipe? 
</details>
Clarified in the Ubuntu-specific instructions.

---

*[drake/doc/from_source.rst, line 107 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBmaXSOd9FrEL-Fnxw:-KKGCENZv29pySw4hxZY:-618708269) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L107)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I am compiling from home, no Matlab. So I tried:
> 
> > `env CXX=g++-4.9 CC=gcc-4.9 cmake path/to/drake-distro -G Ninja -DDISABLE_MATLAB=ON`
> 
> That worked. Maybe worth documenting?
</details>
If you don't have MATLAB, `DISABLE_MATLAB` should be redundant, right? 

Regardless, I don't want this to become a catchall document for our cmake options.

---

*[drake/doc/from_source.rst, line 108 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKA_ygR4x6XdAinJhEE-r2-108:-KKG8CrSLEpmdAciQSxk:762100995) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L108)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Unfortunately, using `-j 1` did not fix the problem.
</details>
Removed `-j` from the doc.

---

*[drake/doc/from_source.rst, line 111 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBjQ1LjEhf08zJQgvK:-KKG8mx6ObeI-5LfPBUf:-120904363) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L111)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Not the most user friendly files I've seen. Is there an equivalent to `make options` that goes into the ccmake interface?
</details>
Yes, `ccmake`.  Added to doc.

---

*[drake/doc/from_source.rst, line 113 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBmco8jSNbn47f1Vtk:-KKGC_ysl9K1WpOTIML4:2087125187) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L113)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Other questions:
> 
> - How do I build from scratch at the drake level? (without building at the superbuild level).
> - I also noticed that from within `drake-distro/drake-build/drake` I can build specific targets with `ninja -j <num_CPUs> target_name`. Worth documenting?
> - The progress is reported on a per-target basis. I can see a list of files being compiled with `ninja -v`. However I get a super lengthy chunk of information with all the compilation flags and paths. Is there a way to ask for a reporting level somewhere in between?  
> - Out-of-source questions:
>     - Could I have a `drake-build/Release` and a `drake-build/Debug`? is that the recommended usage?
>     - Sometimes compilation gets slow when checking out different branches. Would you recommend having a `drake-build/branch_name/[Release|Debug]`?
</details>
(1) and (2) - Done.

(3) - I believe there isn't.

(4) - I've never tried that.  Feel free to look into it and update this doc!

---

*[drake/doc/from_source.rst, line 124 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBbpk2oNzkSEuVZEn3:-KKGCk9QNvVhTiTg-sGQ:-596423098) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L124)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Probably people like @liangfok and me that are switching from make to Ninja might need to install ninja. Of course we won't go to the Ubuntu install instructions to find out the appropriate ninja-build package. Would it be a good idea to reference that prerequisite here? just a link to the updated Ubuntu install instructions should do it I think.
</details>
You can always click the Ninja link and download it.  Installation via apt is not mandatory.  I'd prefer not to pile on cross-references for each platform in this page; it would start to defeat the purpose of having any common instructions at all.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/doc/from_source.rst, line 97 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBf_fjN11dawH2i6K_:-KKJ8vhkQ4gUDYLAz4L8:-1976324695) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L97)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> As @liangfok said and is noted in the paragraph above, the build is out-of-source.
</details>
I see. I was putting it in `drake-distro`. Thanks!

---

*[drake/doc/from_source.rst, line 99 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBcvRhCQzh0QPAK1_j:-KKJ9ADVk2sD4DvbbZx2:-1289920481) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L99)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Clarified in the Ubuntu-specific instructions.
</details>
I see how if you add my suggestion we run the risk of doc going out of sync. Should we add then a link here to the "platform specific instructions"?

---

*[drake/doc/from_source.rst, line 107 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBmaXSOd9FrEL-Fnxw:-KKJ9deoCIJ0e6armDuj:141625952) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L107)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> If you don't have MATLAB, `DISABLE_MATLAB` should be redundant, right? 
> 
> Regardless, I don't want this to become a catchall document for our cmake options.
</details>
Is not redundant. I do have Matlab. However at home I don't have access to the license. Therefore I get a license error at compile time. I agree with you on the not catchall document. However this options is not documented anywhere. 

---

*[drake/doc/from_source.rst, line 111 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBjQ1LjEhf08zJQgvK:-KKJAhPDDUQMBY_AVuBO:765390629) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L111)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Yes, `ccmake`.  Added to doc.
</details>
so we can do `ccmake path/to/drake-distro -G Ninja -DCMAKE_BUILD_TYPE:STRING=[Debug|Release|RelWithDebInfo]` instead of `cmake path/to/drake-distro -G Ninja -DCMAKE_BUILD_TYPE:STRING=[Debug|Release|RelWithDebInfo]`?

---

*[drake/doc/from_source.rst, line 113 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBmco8jSNbn47f1Vtk:-KKJtF6c0fZvfGcEDffX:836695540) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L113)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> (1) and (2) - Done.
> 
> (3) - I believe there isn't.
> 
> (4) - I've never tried that.  Feel free to look into it and update this doc!
</details>
Thanks! I'll keep playing around with it and hopefully I'll have an answer for 4 soon

---

*[drake/doc/from_source.rst, line 124 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBbpk2oNzkSEuVZEn3:-KKJtZ7QH0Av5eILhuwU:468266984) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L124)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> You can always click the Ninja link and download it.  Installation via apt is not mandatory.  I'd prefer not to pile on cross-references for each platform in this page; it would start to defeat the purpose of having any common instructions at all.
</details>
Agree thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Other than the ordering issue mentioned by @bradking above, this documentation <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>.

Perhaps instead of mentioning the ordering issue in this documentation, we should just wait for #2487 to be merged before merging this PR.

---

Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/doc/from_source.rst, line 107 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBmaXSOd9FrEL-Fnxw:-KKJvunyIr0DOkOmpu-O:-506411724) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L107)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Is not redundant. I do have Matlab. However at home I don't have access to the license. Therefore I get a license error at compile time. I agree with you on the not catchall document. However this options is not documented anywhere. 
</details>
I also ran into this have-matlab-but-cannot-access-license-server issue. I'm OK not documenting it because I suspect it's TRI specific.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKJvIdPguDwvJr_33kI:-1008978617)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/doc/from_source.rst, line 69 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKLPE-EQASNiTGFLmbl:-KKLPE-EQASNiTGFLmbm:247045360) ([raw file](https://github.com/robotlocomotion/drake/blob/2f7b3cd2aea5e108faabea345d22b9f90a62f9a8/drake/doc/from_source.rst#L69)):*
> ```reStructuredText
> 	fedora
> 	homebrew
> 	ubuntu
> ```

The `from_source.rst` instructions say to detour and follow the `ubuntu.rst` instructions, which include things like:
```
cd drake-distro
make options
make download-all
```

Are those instructions relevant (or correct) for users who will be using ninja?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I still have some questions but I guess they'll go away once I start using this on a regular basis. <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKOITVR5dW3CqD3DmOO:-320047231)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/doc/from_source.rst, line 99 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBcvRhCQzh0QPAK1_j:-KKPRFiigPus_UFFwDMq:-964580672) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L99)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I see how if you add my suggestion we run the risk of doc going out of sync. Should we add then a link here to the "platform specific instructions"?
</details>
There already is one, just a few lines up :-)

---

*[drake/doc/from_source.rst, line 107 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBmaXSOd9FrEL-Fnxw:-KKPRPhdeFq23kOJi5zj:-323635673) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L107)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I also ran into this have-matlab-but-cannot-access-license-server issue. I'm OK not documenting it because I suspect it's TRI specific.
</details>
OK, if we're going to document it (and I tentatively agree with @amcastro-tri that we should) it's a separate issue.  I could imagine a whole documentation page devoted to CMake options.

---

*[drake/doc/from_source.rst, line 111 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBjQ1LjEhf08zJQgvK:-KKPRbh3eAZFhR9WR1f0:1217436540) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L111)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> so we can do `ccmake path/to/drake-distro -G Ninja -DCMAKE_BUILD_TYPE:STRING=[Debug|Release|RelWithDebInfo]` instead of `cmake path/to/drake-distro -G Ninja -DCMAKE_BUILD_TYPE:STRING=[Debug|Release|RelWithDebInfo]`?
</details>
Yes.  Further clarified the doc.

---

*[drake/doc/from_source.rst, line 69 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKLPE-EQASNiTGFLmbl:-KKPUAnxRkkTGUSSpk0A:745833507) ([raw file](https://github.com/robotlocomotion/drake/blob/2f7b3cd2aea5e108faabea345d22b9f90a62f9a8/drake/doc/from_source.rst#L69)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The `from_source.rst` instructions say to detour and follow the `ubuntu.rst` instructions, which include things like:
> ```
> cd drake-distro
> make options
> make download-all
> ```
> 
> Are those instructions relevant (or correct) for users who will be using ninja?
</details>
Good catch, cleaned that up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/doc/homebrew.rst, line 20 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKRj67t5CkwB_bJR79j:-KKRj67ud7IH_ytn8MHY:-453728444) ([raw file](https://github.com/robotlocomotion/drake/blob/ba92fcc5ebbba06d3dcce6aa47624392b9905546/drake/doc/homebrew.rst#L20)):*
> ```reStructuredText
>     brew upgrade
>     brew install autoconf automake cmake doxygen gcc glib graphviz gtk+ \
>       jpeg libpng libtool mpfr mpich2 ninja numpy python qt qwt swig valgrind \
> ```

BTW I wonder whether this installs version 1.7.1, which is used by CI (see: https://github.com/RobotLocomotion/drake/pull/2555#issuecomment-225952186). How important is it that we use the same version of `ninja` locally as what's used on CI?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/doc/homebrew.rst, line 20 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKRj67t5CkwB_bJR79j:-KKTTJpPMfrf3DA_QozH:1200095742) ([raw file](https://github.com/robotlocomotion/drake/blob/ba92fcc5ebbba06d3dcce6aa47624392b9905546/drake/doc/homebrew.rst#L20)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW I wonder whether this installs version 1.7.1, which is used by CI (see: https://github.com/RobotLocomotion/drake/pull/2555#issuecomment-225952186). How important is it that we use the same version of `ninja` locally as what's used on CI?
</details>
It will install 1.7.1 at present and tends to track the latest version with a small amount of lag. You should probably have the version installed on CI or newer, when that is applicable, but so long as the version is relatively recent, in practice it should be fine. The same could be said for a large number of dependencies on OS X that are installed by Homebrew.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/doc/from_source.rst, line 111 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKBjQ1LjEhf08zJQgvK:-KKTWedokvAvtMTDGWd_:-1993424399) ([raw file](https://github.com/robotlocomotion/drake/blob/25b88843b8c5a01264ec2f91596e56f6806ced76/drake/doc/from_source.rst#L111)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary>
> Yes.  Further clarified the doc.
</details>
Great! thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/doc/from_source.rst, line 106 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKUlIZs070Jz5l7-9AK:-KKUlIZs070Jz5l7-9AL:-910013155) ([raw file](https://github.com/robotlocomotion/drake/blob/ba92fcc5ebbba06d3dcce6aa47624392b9905546/drake/doc/from_source.rst#L106)):*
> ```reStructuredText
>     ninja
> 
> Ninja can rebuild Drake from within the `/drake` subdirectory without
> ```

BTW: nit: \`/drake\` → \`drake-build/drake\`

---

*[drake/doc/homebrew.rst, line 20 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKRj67t5CkwB_bJR79j:-KKUl9Pw7VbMi1sLOm9q:-633812880) ([raw file](https://github.com/robotlocomotion/drake/blob/ba92fcc5ebbba06d3dcce6aa47624392b9905546/drake/doc/homebrew.rst#L20)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary>
> It will install 1.7.1 at present and tends to track the latest version with a small amount of lag. You should probably have the version installed on CI or newer, when that is applicable, but so long as the version is relatively recent, in practice it should be fine. The same could be said for a large number of dependencies on OS X that are installed by Homebrew.
</details>
Great. Thanks for the info!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

liangfok commented:
What's holding up this PR from being merged? If it's my request that `ninja` work on PCs with 16GB of RAM and 1.5GB of swap, I recommend going ahead and merging this PR since that issue is already tracked by https://github.com/RobotLocomotion/drake/issues/2074.

naveenoid commented:
I have a problem with building with gurobi in following these instructions. The build process simply pauses indefinitely at some point. Pressing enter does nothing and neither am I prompted to do anything. Also if a indefinitely paused build is terminated with Ctrl+C, build without the gurobi option does work anymore either. If this is worth a separate issue, I will go do that and not hold up this PR.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKcrqHReXC1a-Pk-5tb:-693226389)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 3 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Added a mention that Gurobi doesn't work with Ninja.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[drake/doc/from_source.rst, line 106 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-KKUlIZs070Jz5l7-9AK:-KKjRax7sePL98SAERNL:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/ba92fcc5ebbba06d3dcce6aa47624392b9905546/drake/doc/from_source.rst#L106)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: nit: \`/drake\` → \`drake-build/drake\`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2555#-:-KKjROQVyYfPbgzBeuzM:1466796700)*
<!-- Sent from Reviewable.io -->
